framework/src/common/errors.rs

```diff
@@ -1,98 +1,145 @@
+    FailedToDropPackets,
 
-        ConfigurationError(description: String) {
-            description("Configuration error")
-            display("Configuration error: {}", description)
-        }
+    #[fail(display = "Failed to deallocate memory")]
+    FailedDeallocation,
 
-        NoRunningSchedulerOnCore(core: i32) {
-            description("No scheduler running on core")
-            display("No scheduler running on core {}", core)
-        }
+    #[fail(display = "Failed to initialize port: {}", _0)]
+    FailedToInitializePort(i32),
 
-        FailedToInsertHeader {
-            description("Failed to insert header into packet")
-            display("Failed to insert header into packet")
-        }
+    #[fail(display = "Invalid queue request")]
+    BadQueue,
 
-        FailedToSwapHeader(new_header: String) {
-            description("Failed to swap-in new header in packet")
-                display("Failed to swap-in new header - {} - in packet", new_header)
-        }
+    #[fail(display = "Cannot send data out port")]
+    CannotSend,
 
-        FailedToRemoveHeader {
-            description("Failed to remove header from packet")
-                display("Failed to remove header from packet")
-        }
+    #[fail(display = "Cannot find device: {}", _0)]
+    BadDev(String),
 
-        FailedToParseMacAddress(s: String) {
-            description("Failed to parse MAC address")
-                display("Failed to parse MAC address: '{}'", s)
-        }
+    #[fail(display = "Bad vdev specification: {}", _0)]
+    BadVdev(String),
+
+    #[fail(display = "Bad TX queue {} for port {}", _0, _1)]
+    BadTxQueue(i32, i32),
+
+    #[fail(display = "Bad RX queue {} for port {}", _0, _1)]
+    BadRxQueue(i32, i32),
+
+    #[fail(display = "Attempt to access bad packet offset {}", _0)]
+    BadOffset(usize),
+
+    #[fail(display = "Metadata is too large")]
+    MetadataTooLarge,
+
+    #[fail(display = "Could not allocate ring")]
+    RingAllocationFailure,
+
+    #[fail(display = "Address of second copy of ring does not match expected address")]
+    RingDuplicationFailure,
+
+    #[fail(display = "Bad ring size {}, must be a power of 2", _0)]
+    InvalidRingSize(usize),
+
+    #[fail(display = "Configuration error: {}", _0)]
+    ConfigurationError(String),
+
+    #[fail(display = "No scheduler running on core {}", _0)]
+    NoRunningSchedulerOnCore(i32),
+
+    #[fail(display = "Failed to insert header into packet")]
+    FailedToInsertHeader,
+
+    #[fail(
+        display = "Failed to swap-in new header - {} - in packet, new_header",
+        _0
+    )]
+    FailedToSwapHeader(String),
+
+    #[fail(display = "Failed to remove header from packet")]
+    FailedToRemoveHeader,
+
+    #[fail(display = "Failed to parse MAC address: '{}'", _0)]
+    FailedToParseMacAddress(String),
+
+    #[fail(display = "_")]
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+#[macro_export]
+macro_rules! error_chain {
```

Comment:
@drunkirishcoder added these macros as the debug print on an `failure::Error` type does not provide the nice display format along with the backtrace, etc... So, to make these useful anyhow, I added to ability to log causes too if you want the entire chain.